---
title: "scRNAseq_analysis"
author: "Lasse Meyer"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Software requirements

```{r lib, message=FALSE,warning=FALSE}
#Data containers
library(SpatialExperiment)
library(SingleCellExperiment)

#Data processing
library(scuttle)
library(scater)
library(scran)
library(imcRtools)

#Data visualization
library(cytomapper)
library(dittoSeq)
library(pheatmap)
library(cowplot)
library(ComplexHeatmap)
library(circlize)

#Utilities
library(tidyverse)
library(openxlsx)
library(stringr)
library(patchwork)
library(parallel)
library(BiocParallel)
library(BiocNeighbors)
library(BiocSingular)
library(tiff)
library(ggrepel)
library(EBImage)
library(viridis)
library(mclust)
library(Rphenograph)
library(igraph)
library(RColorBrewer)
library(foreach)
library(clValid)
library(bluster)
library(psych)
library(corrplot)
library(psych)

#Batch correction
library(batchelor)
```

# scRNAseq analysis 

Data from GSE161529 - 8 TNBC patients (4 BRCA-/+) - Pal et al. (2021, EMBO Journal)

## 1. Load and prepare data 

```{r}
library(DropletUtils)

dir_data <- list.files("/mnt/tnbc_volume/GSE161529", full.names = TRUE)[2:9]

sce <- read10xCounts(dir_data)
sce
```


### Add available metadata and process sce

```{r}
library(readxl)

meta <- read_excel("/mnt/tnbc_volume/GSE161529/metadata_TNBC_GSE161529.xlsx") %>% as.data.frame

sce$sample_id <- str_split(sce$Sample, "/", simplify = TRUE)[,5]
sce$Sample <- NULL

## Add other relevant metadata based on sample_id
for (entry in colnames(meta)[!colnames(meta) %in% c("sample_id")]) {
  colData(sce)[,entry] <- meta[,entry][match(colData(sce)[,"sample_id"],meta[,"sample_id"])]
}

## Generate cell identifiers 
rownames(colData(sce)) <- sce$cell_id <- paste0(sce$PID,"_",seq(length(sce$Barcode)))

# Number of cells per PID 
colData(sce) %>% as.data.frame %>% group_by(PID) %>% count()

## Change rownames 
rownames(sce) <- rowData(sce)$Symbol

sce
```

Save sce before QC 

```{r}
saveRDS(sce, "/mnt/tnbc_volume/GSE161529/data_analysis/sce_scRNA.rds")
```

## 2. Quality control 

```{r}
# Identifying the mitochondrial transcripts in our SingleCellExperiment.
is.mito <- grepl("^MT",rownames(sce))

library(scuttle)
df <- perCellQCMetrics(sce, subsets=list(Mito=is.mito))

summary(df$sum) #total count for each cell
summary(df$detected) #the number of detected genes
summary(df$subsets_Mito_percent) #percentage of reads mapped to mitochondrial transcripts

# Add QC to sce
sce <- addPerCellQCMetrics(sce, subsets=list(Mito=is.mito))
```



```{r}
# Remove based on fixed threshold 
## 500 detected genes - max. 20 % of mitochondrial reads

qc_detected <- df$detected < 500
qc_mito <- df$subsets_Mito_percent > 20

# Adaptive thresholds (based on median absolute deviations)
reasons <- perCellQCFilters(df, 
    sub.fields=c("subsets_Mito_percent"))

discard <- qc_detected | qc_mito | reasons$discard

summary(discard)

# Overview
DataFrame(qc_mito =sum(qc_mito), qc_detected = sum(qc_detected),Total=sum(discard))

# Subset sce
sce <- sce[,!discard]
```

Save sce after QC 

```{r}
saveRDS(sce, "/mnt/tnbc_volume/GSE161529/data_analysis/sce_scRNA_QC.rds")
```

## 3. Normalization 

```{r}
set.seed(22)
clust <- quickCluster(sce) 
sce <- computeSumFactors(sce, cluster=clust, min.mean=0.1)
sce <- logNormCounts(sce)
assayNames(sce)
```

## 4. Feature selection 

```{r}
var_sce <- modelGeneVar(sce)
hvg_sce <- getTopHVGs(var_sce, prop=0.1)
hvg <- str(hvg_sce)

# Check whether cytokeratin 5/7 are part of hvgs
summary(hvg_sce %in% c("KRT5","KRT7"))

# Add to sce
rowSubset(sce, "HVG") <- getTopHVGs(var_sce, prop=0.1)
colnames(rowData(sce))
```


## 5. Dimensionality reduction

```{r}
library(scater)
set.seed(22)

#Use 30 PCs for denoising 
sce <- runPCA(sce, ncomponents=30, subset_row=hvg_sce)<

#Run UMAP and TSNE on selected PCs
sce <- runUMAP(sce, subset_row = hvg_sce, n_neighbors = 30, dimred = "PCA", exprs_values = "logcounts", external_neighbors=TRUE, BPPARAM = MulticoreParam())
sce <- runTSNE(sce, subset_row = hvg_sce, n_neighbors = 30, dimred = "PCA", exprs_values = "logcounts", external_neighbors=TRUE, BPPARAM = MulticoreParam())

# Visualize 
plotReducedDim(sce, dimred="PCA", colour_by="PID")
plotReducedDim(sce, dimred="UMAP", colour_by="PID")
plotReducedDim(sce, dimred="TSNE", colour_by="PID")

## Strong sample-specific effects are present (batch-correction needed)
```

# 6. Batch correction - fastMNN 

```{r}
library(batchelor)
set.seed(22)
out <- fastMNN(sce, batch = sce$sample_id,
               auto.merge = TRUE,
               subset.row = rowData(sce)$HVG,
               assay.type = "logcounts")

# Transfer the correction results to the main sce object
reducedDim(sce, "fastMNN") <- reducedDim(out, "corrected")

# Visualize 
sce <- runTSNE(sce, subset_row = rowData(sce)$HVG, dimred = "fastMNN", name = "TSNE_MNN",BPPARAM = MulticoreParam())
plotReducedDim(sce, dimred="TSNE_MNN", colour_by="PID")
```

# 7. Distinguish epithelial and non-epithelial cells 

## Clustering

```{r}
graph_clust <- clusterCells(x = sce, use.dimred = "fastMNN",
                      BLUSPARAM = NNGraphParam(k=30,
                                               cluster.fun ="louvain",
                                               type="jaccard",
                                               BPPARAM=MulticoreParam(workers = 20)))
sce$cluster_30 <- graph_clust                   
length(unique(sce$cluster_30))

plotReducedDim(sce, dimred="TSNE_MNN", colour_by="cluster_30", text_by = "cluster_30")
```

## GMM for EPCAM/KRT5/KRT8

We will use a two component gaussian mixture model to distinguish between tumor (EPCAM+) and stromal (EPCAM-) cells. 

```{r GMM_CK, message=FALSE,warning=FALSE}
## Gaussian mixture model to distinguish tumor/stroma based on EPCAM expression
GMM_epi <- Mclust(assay(sce, "logcounts")["EPCAM",], G = 2)

GMM_df <- data.frame(class_epi = ifelse(GMM_epi$classification == 2, #& GMM_epi$uncertainty == 0, 
                                        "EPCAM+",
                                        "EPCAM-"), 
                     EPCAM_exprs = assay(sce, "logcounts")["EPCAM",],
                     uncertainty = GMM_epi$uncertainty)

#Visualize GMM
ggplot(GMM_df, aes(x = EPCAM_exprs, fill = class_epi))+
  geom_density()+
  theme_classic(base_size = 15)+
  theme(axis.title.x = element_blank())+
  scale_fill_brewer(palette = "Set1")
  
#Add GMM epi class to SCE
sce$GMM_EPCAM <- GMM_df$class_epi

plotReducedDim(sce, dimred="TSNE_MNN", colour_by="GMM_EPCAM")

## Gaussian mixture model to distinguish tumor/stroma based on KRT5 (basal) expression
GMM_epi <- Mclust(assay(sce, "logcounts")["KRT5",], G = 2)

GMM_df <- data.frame(class_epi = ifelse(GMM_epi$classification == 2, #& GMM_epi$uncertainty == 0, 
                                        "KRT5+",
                                        "KRT5-"), 
                     KRT5_exprs = assay(sce, "logcounts")["KRT5",],
                     uncertainty = GMM_epi$uncertainty)

#Visualize GMM
ggplot(GMM_df, aes(x = KRT5_exprs, fill = class_epi))+
  geom_density()+
  theme_classic(base_size = 15)+
  theme(axis.title.x = element_blank())+
  scale_fill_brewer(palette = "Set1")
  
#Add GMM epi class to SCE
sce$GMM_KRT5 <- GMM_df$class_epi

plotReducedDim(sce, dimred="TSNE_MNN", colour_by="GMM_KRT5")

## Gaussian mixture model to distinguish tumor/stroma based on KRT8 (luminal) expression
GMM_epi <- Mclust(assay(sce, "logcounts")["KRT8",], G = 2)

GMM_df <- data.frame(class_epi = ifelse(GMM_epi$classification == 2, #& GMM_epi$uncertainty == 0, 
                                        "KRT8+",
                                        "KRT8-"), 
                     KRT8_exprs = assay(sce, "logcounts")["KRT8",],
                     uncertainty = GMM_epi$uncertainty)

#Visualize GMM
ggplot(GMM_df, aes(x = KRT8_exprs, fill = class_epi))+
  geom_density()+
  theme_classic(base_size = 15)+
  theme(axis.title.x = element_blank())+
  scale_fill_brewer(palette = "Set1")
  
#Add GMM epi class to SCE
sce$GMM_KRT8 <- GMM_df$class_epi

plotReducedDim(sce, dimred="TSNE_MNN", colour_by="GMM_KRT8")

```

## Define and visualize analysis categories 

Heatmap with clustering and GMM information 

```{r}

sce <- readRDS("/mnt/tnbc_volume/GSE161529/data_analysis/sce_scRNA_anacat.rds")

### 1. Heatmap body ###

# Min-Max [0,1] normalization of arcsinh-transformed counts
assay(sce, "min_max") <- assay(sce, "logcounts") - rowMins(assay(sce, "logcounts"))
assay(sce, "min_max") <- assay(sce, "min_max") / (rowMaxs(assay(sce, "logcounts")) - rowMins(assay(sce, "logcounts")))

# Aggregrate across cells with the mean
mean_sce <- aggregateAcrossCells(sce, ids = sce$cluster_30,
                                 statistics = "mean",
                                 use.assay.type = "logcounts",
                                 subset.row = rowData(sce)$Symbol %in% c("EPCAM"))

# Heatmap body color 
col_exprs <- colorRamp2(c(0,1,2,3), c("#440154FF", "#31688EFF", "#35B779FF", "#FDE725FF"))


### 2. Heatmap annotation ###

### 2.1 Metadata features

#Number of cells per cluster 
n_cells <- colData(mean_sce) %>% as.data.frame %>% select(cluster_30, ncells)

#Define analysis categories based on GMM 
nr_EPCAM <- colData(sce) %>% as.data.frame %>% select(cluster_30, GMM_EPCAM) %>% table()
fr_EPCAM <- as.data.frame(nr_EPCAM) %>% pivot_wider(names_from = GMM_EPCAM, values_from = Freq) %>% column_to_rownames("cluster_30")
fr_EPCAM <- fr_EPCAM %>% transmute(fr_EPCAM = `EPCAM+`/rowSums(fr_EPCAM)) %>% as.data.frame()

nr_KRT5 <- colData(sce) %>% as.data.frame %>% select(cluster_30, GMM_KRT5) %>% table()
fr_KRT5 <- as.data.frame(nr_KRT5) %>% pivot_wider(names_from = GMM_KRT5, values_from = Freq) %>% column_to_rownames("cluster_30")
fr_KRT5 <- fr_KRT5 %>% transmute(fr_KRT5 = `KRT5+`/rowSums(fr_KRT5)) %>% as.data.frame()

nr_KRT8 <- colData(sce) %>% as.data.frame %>% select(cluster_30, GMM_KRT8) %>% table()
fr_KRT8 <- as.data.frame(nr_KRT8) %>% pivot_wider(names_from = GMM_KRT8, values_from = Freq) %>% column_to_rownames("cluster_30")
fr_KRT8 <- fr_KRT8 %>% transmute(fr_KRT8 = `KRT8+`/rowSums(fr_KRT8)) %>% as.data.frame()

anno_cluscat <- cbind(fr_EPCAM,fr_KRT5,fr_KRT8) 
anno_cluscat <- anno_cluscat %>% mutate(cluster_30 = rownames(anno_cluscat))
anno_cluscat$max <- anno_cluscat %>% pivot_longer(names_to = "GMM", values_to = "fra",cols = 1:3) %>% group_by(cluster_30) %>% mutate(max = max(fra)) %>% select(cluster_30, max) %>% unique() %>% pull(max)

# Define analysis_cat 
anno_cluscat <- anno_cluscat %>% 
  mutate(cluster_category = ifelse(fr_EPCAM >= 0.3,"Epithelial","Non-Epithelial"))

# Add to sce
sce$analysis_cat <- anno_cluscat$cluster_category[match(sce$cluster_30, anno_cluscat$cluster_30)]

analysis_cat <- setNames(c("#E41A1C", "#377EB8"),sort(unique(sce$analysis_cat)))                                  
metadata(sce)$colors$analysis_cat <- analysis_cat 

# Visualize on TSNE
plotReducedDim(sce, dimred="TSNE_MNN", colour_by="analysis_cat")+
  scale_color_manual(values = metadata(sce)$colors$analysis_cat)

# Visualize as Heatmap 

h <- Heatmap(t(assay(mean_sce, "logcounts")),
        col = col_exprs,
        name= "logcounts",
        show_row_names = T, 
        show_column_names =  T, 
        clustering_method_rows = "ward.D2",
        clustering_method_columns = "ward.D2")+
  rowAnnotation(clus_cat = anno_simple(anno_cluscat$cluster_category, col = metadata(sce)$colors$analysis_cat))
draw(h)
```

Save sce after QC 

```{r}
saveRDS(sce, "/mnt/tnbc_volume/GSE161529/data_analysis/sce_scRNA_anacat.rds")
```
