---
title: "scRNAseq_analysis"
author: "Lasse Meyer"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Software requirements

```{r lib, message=FALSE,warning=FALSE}
#Data containers
library(SpatialExperiment)
library(SingleCellExperiment)

#Data processing
library(scuttle)
library(scater)
library(scran)
library(imcRtools)

#Data visualization
library(cytomapper)
library(dittoSeq)
library(pheatmap)
library(cowplot)
library(ComplexHeatmap)
library(circlize)

#Utilities
library(tidyverse)
library(openxlsx)
library(stringr)
library(patchwork)
library(parallel)
library(BiocParallel)
library(BiocNeighbors)
library(BiocSingular)
library(tiff)
library(ggrepel)
library(EBImage)
library(viridis)
library(mclust)
library(Rphenograph)
library(igraph)
library(RColorBrewer)
library(foreach)
library(clValid)
library(bluster)
library(psych)
library(corrplot)
library(psych)

#Batch correction
library(batchelor)
```

# scRNAseq analysis 

Data from GSE161529 - 8 TNBC patients (4 BRCA-/+) - Pal et al. (2021, EMBO Journal)

## 1. Load and prepare data 

```{r}
library(DropletUtils)

dir_data <- list.files("/mnt/tnbc_volume/GSE161529", full.names = TRUE)[1:8]


sce <- read10xCounts(dir_data)
sce
```


### Add available metadata and process sce

```{r}
library(readxl)

meta <- read_excel("/mnt/tnbc_volume/GSE161529/metadata_TNBC_GSE161529.xlsx") %>% as.data.frame

sce$sample_id <- str_split(sce$Sample, "/", simplify = TRUE)[,5]
sce$Sample <- NULL

## Add other relevant metadata based on sample_id
for (entry in colnames(meta)[!colnames(meta) %in% c("sample_id")]) {
  colData(sce)[,entry] <- meta[,entry][match(colData(sce)[,"sample_id"],meta[,"sample_id"])]
}

## Generate cell identifiers 
rownames(colData(sce)) <- sce$cell_id <- paste0(sce$PID,"_",seq(length(sce$Barcode)))

# Number of cells per PID 
colData(sce) %>% as.data.frame %>% group_by(PID) %>% count()

## Change rownames 
rownames(sce) <- rowData(sce)$Symbol

sce
```

Save sce before QC 

```{r}
saveRDS(sce, "/mnt/tnbc_volume/GSE161529/data_analysis/sce_scRNA.rds")
```

## 2. Quality control 

```{r}
# Identifying the mitochondrial transcripts in our SingleCellExperiment.
is.mito <- grepl("^MT",rownames(sce))

library(scuttle)
df <- perCellQCMetrics(sce, subsets=list(Mito=is.mito))

summary(df$sum) #total count for each cell
summary(df$detected) #the number of detected genes
summary(df$subsets_Mito_percent) #percentage of reads mapped to mitochondrial transcripts

# Add QC to sce
sce <- addPerCellQCMetrics(sce, subsets=list(Mito=is.mito))
```



```{r}
# Remove based on fixed threshold 
## 500 detected genes - max. 20 % of mitochondrial reads

qc_detected <- df$detected < 500
qc_mito <- df$subsets_Mito_percent > 20

# Overview
DataFrame(qc_mito =sum(qc_mito), qc_detected = sum(qc_detected),Total=sum(discard))

# Adaptive thresholds (based on median absolute deviations)
reasons <- perCellQCFilters(df, 
    sub.fields=c("subsets_Mito_percent"))

discard <- qc_sum | qc_mito | reasons$discard

summary(discard)

# Subset sce
sce <- sce[,!discard]
```

Save sce after QC 

```{r}
saveRDS(sce, "/mnt/tnbc_volume/GSE161529/data_analysis/sce_scRNA_QC.rds")
```

## 3. Normalization 

```{r}
set.seed(22)
clust <- quickCluster(sce) 
sce <- computeSumFactors(sce, cluster=clust, min.mean=0.1)
sce <- logNormCounts(sce)
assayNames(sce)
```

## 4. Feature selection 

```{r}
var_sce <- modelGeneVar(sce)
hvg_sce <- getTopHVGs(var_sce, prop=0.1)
hvg <- str(hvg_sce)

# Check whether cytokeratin 5/7 are part of hvgs
summary(hvg_sce %in% c("KRT5","KRT7"))

# Add to sce
rowSubset(sce, "HVG") <- getTopHVGs(var_sce, prop=0.1)
colnames(rowData(sce))
```


## 5. Dimensionality reduction

```{r}
library(scater)
set.seed(22)

#Use 30 PCs for denoising 
sce <- runPCA(sce, ncomponents=30, subset_row=hvg_sce)
sce <- runPCA(sce, ncomponents=30, subset_row=hvg_sce)

#Run UMAP and TSNE on selected PCs
sce <- runUMAP(sce, subset_row = hvg_sce, n_neighbors = 30, dimred = "PCA", exprs_values = "logcounts", external_neighbors=TRUE, BPPARAM = MulticoreParam())
sce <- runTSNE(sce, subset_row = hvg_sce, n_neighbors = 30, dimred = "PCA", exprs_values = "logcounts", external_neighbors=TRUE, BPPARAM = MulticoreParam())

# Visualize 
plotReducedDim(sce, dimred="PCA", colour_by="PID")
plotReducedDim(sce, dimred="UMAP", colour_by="PID")
plotReducedDim(sce, dimred="TSNE", colour_by="PID")

## Strong sample-specific effects are present (batch-correction needed)
```

# 6. Batch correction - fastMNN 

```{r}
library(batchelor)
set.seed(22)
out <- fastMNN(sce, batch = sce$sample_id,
               auto.merge = TRUE,
               subset.row = rowData(sce)$HVG,
               assay.type = "logcounts")

# Transfer the correction results to the main sce object
reducedDim(sce, "fastMNN") <- reducedDim(out, "corrected")

# Visualize 
sce <- runTSNE(sce, subset_row = rowData(sce)$HVG, dimred = "fastMNN", name = "TSNE_MNN",BPPARAM = MulticoreParam())
plotReducedDim(sce, dimred="TSNE_MNN", colour_by="cluster_70")
```

# 7. Graph-based clustering 

## First step: Distinguish tumor and non-tumor cells 

# Clustering 
```{r}
graph_clust <- clusterCells(x = sce, use.dimred = "TSNE_MNN",
                      BLUSPARAM = NNGraphParam(k=30,
                                               cluster.fun ="louvain",
                                               type="jaccard",
                                               BPPARAM=MulticoreParam(workers = 20)))
sce$cluster_30 <- graph_clust                   

graph_clust <- clusterCells(x = sce, use.dimred = "TSNE_MNN",
                      BLUSPARAM = NNGraphParam(k=50,
                                               cluster.fun ="louvain",
                                               type="jaccard",
                                               BPPARAM=MulticoreParam(workers = 20)))
                      
sce$cluster_50 <- graph_clust

graph_clust <- clusterCells(x = sce, use.dimred = "TSNE_MNN",
                      BLUSPARAM = NNGraphParam(k=70,
                                               cluster.fun ="louvain",
                                               type="jaccard",
                                               BPPARAM=MulticoreParam(workers = 20)))
                      
sce$cluster_70 <- graph_clust
```

# GMM for EPCAM

```{r}
saveRDS(sce, "/mnt/tnbc_volume/GSE161529/data_analysis/sce_scRNA_clus.rds")
```


# 8. Downstream analysis 

```{r}
plotReducedDim(sce, dimred = "TSNE_MNN", 
               colour_by = "KRT5", 
               by_exprs_values = "logcounts",
               point_size = 0.5)

plotReducedDim(sce, dimred = "TSNE_MNN", 
               colour_by = "KRT7", 
               by_exprs_values = "logcounts",
               point_size = 0.5)

plotReducedDim(sce, dimred = "TSNE_MNN", 
               colour_by = "KRT8", 
               by_exprs_values = "logcounts",
               point_size = 0.5)

plotReducedDim(sce, dimred = "TSNE_MNN", 
               colour_by = "KRT18", 
               by_exprs_values = "logcounts",
               point_size = 0.5)

cur_dat <- t(assay(sce,"logcounts")) %>% as.data.frame %>% select("KRT5","KRT7")

dittoScatterPlot(
    object = sce,
    x.var = "KRT7", y.var = "KRT5",
    color.var = "PID")


summary(rownames(sce) == "KRT5")


ggplot(cur_dat, aes(x=KRT7, y=KRT5)) +
  geom_point(alpha = 0.5)+
  #stat_density_2d(aes(fill = ..level..), geom="polygon", contour_var = "count")+
  theme_classic()

ggplot(cur_dat, aes(x=KRT7, y=KRT5)) +
  geom_point(alpha = 0.5)+
  stat_density_2d(aes(fill = ..level..), geom="polygon", contour_var = "count")+
  theme_classic()
```

