---
title: "TCGA_BRCA_RNAseq_analysis"
author: "Lasse Meyer"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# TCGA BRCA dataset analysis 

## Load data 

```{r}
library(SummarizedExperiment)
library(TCGAbiolinks)

query.exp <- GDCquery(
    project = "TCGA-BRCA", 
    data.category = "Transcriptome Profiling",
    data.type = "Gene Expression Quantification", 
    workflow.type = "STAR - Counts",
    sample.type = c("Primary Tumor")
)

GDCdownload(
    query = query.exp,
    files.per.chunk = 100
)

brca.exp <- GDCprepare(
    query = query.exp, 
    save = TRUE, 
    save.filename = "brcaExp.rda"
)
```

## Add clinical/survival info

```{r}
# Clean object
brca.clean <- brca.exp
colData(brca.clean) <- NULL
brca.clean$PID <- brca.exp$PID

# 1. Add clinical information 
## Cleaned clinical information from UCSC Xena
information.subtype.xena <- read_delim("/mnt/rcc_volume/TNBC/data_analysis/03_data/BRCA_clinicalinfo.txt")
information.subtype.xena <- information.subtype.xena %>% as.data.frame

## Add relevant data based on PID
for (entry in colnames(information.subtype.xena)[colnames(information.subtype.xena) %in% c("ER_Status_nature2012","PR_Status_nature2012","HER2_Final_Status_nature2012","Metastasis_nature2012","Tumor_nature2012","Node_nature2012","PAM50Call_RNAseq","PAM50_mRNA_nature2012", "Age_at_Initial_Pathologic_Diagnosis_nature2012")]) {
  colData(brca.clean)[,entry] <- information.subtype.xena[,entry][match(colData(brca.clean)[,"PID"],information.subtype.xena$patient_id)]
}

# 2. Add survival information 
## Cleaned survival information from UCSC Xena
information.survival_xena <- read_delim("/mnt/rcc_volume/TNBC/data_analysis/03_data/BRCA_survival_Xena.txt")
information.survival_xena$PID <- str_split(information.survival_xena$sample,"-",simplify = TRUE)[,3]
information.survival_xena <- information.survival_xena %>% as.data.frame

for (entry in colnames(information.survival_xena)[!colnames(information.survival_xena) %in% c("sample","_PATIENT","Redaction","PID")]) {
  colData(brca.clean)[,entry] <- information.survival_xena[,entry][match(colData(brca.clean)[,"PID"],information.survival_xena$PID)]
}

colData(brca.clean)

# # Subtype information from TCGA
# infomation.subtype <- TCGAquery_subtype(tumor = "BRCA")
# # Survival information from TCGA
# information.clinical <- GDCquery_clinic(project = "TCGA-BRCA",type = "clinical") 
```

## Filter dataset for TNBC and basal-like

```{r}
brca.clean$TNBC <- ifelse(brca.clean$ER_Status_nature2012 == "Negative" & brca.clean$PR_Status_nature2012 == "Negative" & brca.clean$HER2_Final_Status_nature2012 == "Negative", "TNBC", "Not_TNBC")
brca.clean$TNBC[is.na(brca.clean$TNBC)] <- "Undefined"

# TNBC 
tnbc.clean <- brca.clean[,brca.clean$TNBC == "TNBC"]
tnbc.clean # 122 patients

# Basal
brca.clean$PAM50Call_RNAseq[is.na(brca.clean$PAM50Call_RNAseq)] <- "Undefined"
basal.clean <- brca.clean[,brca.clean$PAM50Call_RNAseq == "Basal"]
basal.clean # 143 patients
```

```{r}
saveRDS(brca.clean,"/mnt/rcc_volume/TNBC/data_analysis/TCGA_BRCA_cleaned_full.rds")

brca.clean <- readRDS("/mnt/rcc_volume/TNBC/data_analysis/TCGA_BRCA_cleaned_full.rds")
```

## DESeq2 processing

### TCGA BRCA - full dataset 
1106 patients 

```{r}
library(DESeq2)
brca.clean$PAM50Call_RNAseq[is.na(brca.clean$PAM50Call_RNAseq)] <- "Undefined"

# Create DESeqDataset
ddsSE <- DESeqDataSet(brca.clean, design = ~ PAM50Call_RNAseq)

ddsSE

# DESeq analysis includes count normalization based on size factors
res <- DESeq(ddsSE)
rownames(res) <- rowData(res)$gene_name

DESeq2::plotCounts(res, gene = "KRT7", intgroup = "PAM50Call_RNAseq")
DESeq2::plotCounts(res, gene = "KRT5", intgroup = "PAM50Call_RNAseq")

counts(res, normalized = TRUE) # Access normalized counts 

# Variance-stabilizing transformation
vsd <- DESeq2::vst(res)

# Access normalized and stabilized counts
mat <- assay(vsd)
```

## CK5/7 Signature 

GMM on size-factor normalized (NOT VST) counts

```{r}
# 1. CK5
GMM_CK5 <- Mclust(counts(res, normalized = TRUE)["KRT5",], G = 2)

GMM_df <- data.frame(class_epi = ifelse(GMM_CK5$classification == 2, 
                                        "KRT5+",
                                        "KRT5-"), 
                     KRT5_exprs = assay(vsd)["KRT5",],
                     uncertainty = GMM_CK5$uncertainty)

#Visualize GMM
ggplot(GMM_df, aes(x = KRT5_exprs, fill = class_epi))+
  geom_density()+
  geom_rug(aes(color = class_epi))+
  theme_classic(base_size = 15)+
  theme(axis.title.x = element_blank())+
  scale_fill_brewer(palette = "Set1")+
  scale_color_brewer(palette = "Set1")

#Add GMM epi class to object
vsd$GMM_KRT5 <- GMM_df$class_epi

# 2. CK7
GMM_CK7 <- Mclust(counts(res, normalized = TRUE)["KRT7",], G = 2)

GMM_df <- data.frame(class_epi = ifelse(GMM_CK7$classification == 2, 
                                        "KRT7+",
                                        "KRT7-"), 
                     KRT7_exprs = assay(vsd)["KRT7",],
                     uncertainty = GMM_CK7$uncertainty)

#Visualize GMM
ggplot(GMM_df, aes(x = KRT7_exprs, fill = class_epi))+
  geom_density()+
  geom_rug(aes(color = class_epi))+
  theme_classic(base_size = 17)+
  theme(axis.title.x = element_blank())+
  scale_fill_brewer(palette = "Set1")+
  scale_color_brewer(palette = "Set1")

#Add GMM epi class to object
vsd$GMM_KRT7 <- GMM_df$class_epi

## Combine KRT5/7 GMM scores 
vsd$KRT5_7_score <- ifelse(vsd$GMM_KRT5 == "KRT5+" & vsd$GMM_KRT7 == "KRT7+", "KRT5/7+", 
       ifelse(vsd$GMM_KRT5 == "KRT5+" & vsd$GMM_KRT7 == "KRT7-", "KRT5+",
              ifelse(vsd$GMM_KRT5 == "KRT5-" & vsd$GMM_KRT7 == "KRT7+", "KRT7+","KRT5/7-")))

## Add colors 
metadata(vsd)$colors$KRT57_score <- setNames(brewer.pal(4,"Set1"),unique(vsd$KRT5_7_score))
metadata(vsd)$colors$PAM50 <- setNames(brewer.pal(6,"Paired"),sort(unique(vsd$PAM50Call_RNAseq)))

## Visualize 
plot <- as.data.frame(t(assay(vsd)[rownames(assay(vsd)) %in% c("KRT5","KRT7"),]))
plot$KRT5_7_score <- vsd$KRT5_7_score[match(rownames(plot),colnames(vsd))]
plot$PAM50Call_RNAseq <- vsd$PAM50Call_RNAseq[match(rownames(plot),colnames(vsd))]
plot$TNBC <- vsd$TNBC[match(rownames(plot),colnames(vsd))]

ggplot(plot, aes(x=KRT7, y=KRT5, color = KRT5_7_score)) +
  geom_point(size = 2)+
  #stat_density_2d(aes(fill = ..level..), geom="polygon", contour_var = "density", bins = 15)+
  theme_classic()+
  xlab("CK7 counts")+
  ylab("CK5 counts")+
  scale_color_manual(values = metadata(vsd)$colors$KRT57_score)
  #coord_cartesian(xlim = c(5,20), ylim = c(5,20))

ggplot(plot, aes(x=KRT7, y=KRT5, color = PAM50Call_RNAseq)) +
  geom_point(size = 2)+
  #stat_density_2d(aes(fill = ..level..), geom="polygon", contour_var = "density", bins = 15)+
  theme_classic()+
  xlab("CK7 counts")+
  ylab("CK5 counts")+
  scale_color_manual(values = metadata(vsd)$colors$PAM50)
  #coord_cartesian(xlim = c(5,20), ylim = c(5,20))

ggplot(plot, aes(x=KRT7, y=KRT5, color = TNBC)) +
  geom_point(size = 2)+
  #stat_density_2d(aes(fill = ..level..), geom="polygon", contour_var = "density", bins = 15)+
  theme_classic()+
  xlab("CK7 counts")+
  ylab("CK5 counts")
  #coord_cartesian(xlim = c(5,20), ylim = c(5,20))

colData(vsd) %>% as.data.frame %>% group_by(KRT5_7_score) %>% count() %>% ungroup() %>% mutate(fra = n/sum(n))
```

```{r}
saveRDS(vsd,"/mnt/rcc_volume/TNBC/data_analysis/TCGA_BRCA_DESeq_PAM50_vsd.rds")
saveRDS(res,"/mnt/rcc_volume/TNBC/data_analysis/TCGA_BRCA_DESeq_PAM50_res.rds")
```

## Survival analysis 

```{r}
#Assess survival impact 
library(survival)
library(survminer)

cur_surv <- colData(vsd) %>% as.data.frame() %>% dplyr::select(PID, KRT5_7_score, DFI, DFI.time, OS, OS.time, DSS, DSS.time, PFI, PFI.time) #%>% filter(!KRT5_7_score %in% c("KRT5+","KRT7+"))

#Plot survival curves
fit <- survfit(Surv(DFI.time, DFI) ~ KRT5_7_score, data = cur_surv)
print(fit)

ggsurvplot(fit,
          pval = TRUE,
          risk.table = TRUE, # Add risk table
          risk.table.col = "strata", # Change risk table color by groups
          linetype = "strata", # Change line type by groups
          surv.median.line = "hv", # Specify median survival
          ggtheme = theme_classic(), # Change ggplot2 theme
          #palette = "Set1",
          log.rank.weights = "n", 
          xlim = c(0, 4000))

#Plot survival curves
fit_1 <- survfit(Surv(PFI.time, PFI) ~ KRT5_7_score, data = cur_surv)
print(fit_1)

ggsurvplot(fit_1,
          pval = TRUE,
          risk.table = TRUE, # Add risk table
          risk.table.col = "strata", # Change risk table color by groups
          linetype = "strata", # Change line type by groups
          surv.median.line = "hv", # Specify median survival
          ggtheme = theme_classic(), # Change ggplot2 theme
          #palette = "Set1",
          log.rank.weights = "n",
          xlim = c(0, 4000))

cur_surv <- colData(vsd_1) %>% as.data.frame() %>% dplyr::select(PID, KRT5_7_score, DFI, DFI.time, OS, OS.time, DSS, DSS.time, PFI, PFI.time) %>% filter(!KRT5_7_score %in% c("KRT5+","KRT7+"))

#Plot survival curves
fit <- survfit(Surv(DFI.time, DFI) ~ KRT5_7_score, data = cur_surv)
print(fit)

ggsurvplot(fit,
          pval = TRUE,
          risk.table = TRUE, # Add risk table
          risk.table.col = "strata", # Change risk table color by groups
          linetype = "strata", # Change line type by groups
          surv.median.line = "hv", # Specify median survival
          ggtheme = theme_classic(), # Change ggplot2 theme
          #palette = "Set1",
          log.rank.weights = "n", 
          xlim = c(0, 4000))

#Plot survival curves
fit_1 <- survfit(Surv(PFI.time, PFI) ~ KRT5_7_score, data = cur_surv)
print(fit_1)

ggsurvplot(fit_1,
          pval = TRUE,
          risk.table = TRUE, # Add risk table
          risk.table.col = "strata", # Change risk table color by groups
          linetype = "strata", # Change line type by groups
          surv.median.line = "hv", # Specify median survival
          ggtheme = theme_classic(), # Change ggplot2 theme
          #palette = "Set1",
          log.rank.weights = "n",
          xlim = c(0, 4000))
```

## Cox model 

```{r}
vsd$age <- vsd$Age_at_Initial_Pathologic_Diagnosis_nature2012
```

```{r}
## 1. Disease-free interval 
### Multivariate cox proportional hazard model on multiple covariates (here KRT5/7 groups)
PID_metagroups <- colData(vsd) %>% as.data.frame() %>% dplyr::select(KRT5_7_score,PID) %>% group_by(PID) %>% table() %>% as.data.frame

PID_metagroups <- PID_metagroups %>% group_by(PID) %>% mutate(fra = Freq/sum(Freq))

PID_metagroups <- PID_metagroups %>% pivot_wider(id_cols = "PID",names_from = "KRT5_7_score",values_from = "fra") %>% column_to_rownames("PID")

colnames(PID_metagroups) <- c("KRT57negative","KRT57positive","KRT5positive","KRT7positive")

Cox_df <- PID_metagroups

#add covariates for correction
Cox_df$DFI <- vsd$DFI[match(rownames(Cox_df),vsd$PID)]
Cox_df$DFI.time <- vsd$DFI.time[match(rownames(Cox_df),vsd$PID)]

#Cox_df$PAM50Call_RNAseq <- as.factor(vsd$PAM50Call_RNAseq[match(rownames(Cox_df),vsd$PID)])
Cox_df$age <- as.numeric(vsd$age[match(rownames(Cox_df),vsd$PID)])
Cox_df$pT_simple <- as.factor(vsd$Tumor_nature2012[match(rownames(Cox_df),vsd$PID)])
Cox_df$pN_simple <- as.factor(vsd$Node_nature2012[match(rownames(Cox_df),vsd$PID)])
Cox_df$pM_simple <- as.factor(vsd$Metastasis_nature2012[match(rownames(Cox_df),vsd$PID)])

#Based on proportions
covariates <- Cox_df %>% select(-DFI,-DFI.time, -age, -pT_simple, -pN_simple, -pM_simple, - KRT57negative) %>% colnames

multi_formula <- as.formula(paste("Surv(DFI.time, DFI) ~ age + pT_simple + pN_simple + pM_simple +", paste(covariates, collapse = " + "), "+ KRT57negative"))

multi_model <- coxph(multi_formula, data = Cox_df)

multi_res <- summary(multi_model)

plot_df_prop <- data.frame(group = rownames(multi_res$coefficients),
                      p_value = signif(multi_res$coefficients[,"Pr(>|z|)"],2),
                      HR = signif(multi_res$coefficients[,"exp(coef)"],2),
                      CI_low = signif(multi_res$conf.int[,"lower .95"],2),
                      CI_high = signif(multi_res$conf.int[,"upper .95"],2),
                      wald_test = signif(multi_res$waldtest["pvalue"],2))


#Plot hazard ratios
hr_prop <- ggplot(plot_df_prop, aes(x = reorder(group,desc(HR)), y = HR, color = p_value < 0.05))+
geom_point(size = 4)+
geom_errorbar(aes(ymax = CI_high, ymin = CI_low))+
coord_cartesian(ylim=c(0,30))+
geom_hline(yintercept = 1,color = "black",linetype = "dashed")+
scale_color_brewer(palette = "Set1")+
xlab("Covariates")+
ylab("HR for PID groups")+
theme_classic()+
theme(axis.text.x = element_text(angle = 90))

hr_prop
```

````{r}
## 2. Progression-free interval 
PID_metagroups <- colData(vsd) %>% as.data.frame() %>% select(KRT5_7_score,PID) %>% group_by(PID) %>% table() %>% as.data.frame

PID_metagroups <- PID_metagroups %>% group_by(PID) %>% mutate(fra = Freq/sum(Freq))

PID_metagroups <- PID_metagroups %>% pivot_wider(id_cols = "PID",names_from = "KRT5_7_score",values_from = "fra") %>% column_to_rownames("PID")

colnames(PID_metagroups) <- c("KRT57negative","KRT57positive","KRT5positive","KRT7positive")

## 1. Based on SCP assignments
Cox_df <- PID_metagroups

#add covariates for correction
Cox_df$PFI <- vsd$PFI[match(rownames(Cox_df),vsd$PID)]
Cox_df$PFI.time <- vsd$PFI.time[match(rownames(Cox_df),vsd$PID)]

Cox_df$age <- as.numeric(vsd$age[match(rownames(Cox_df),vsd$PID)])
Cox_df$pT_simple <- as.factor(vsd$Tumor_nature2012[match(rownames(Cox_df),vsd$PID)])
Cox_df$pN_simple <- as.factor(vsd$Node_nature2012[match(rownames(Cox_df),vsd$PID)])
Cox_df$pM_simple <- as.factor(vsd$Metastasis_nature2012[match(rownames(Cox_df),vsd$PID)])

#Based on proportions
covariates <- Cox_df %>% select(-PFI,-PFI.time, -age, -pT_simple, -pN_simple, -pM_simple, - KRT57negative) %>% colnames

multi_formula <- as.formula(paste("Surv(PFI.time, PFI) ~ age + pT_simple + pN_simple + pM_simple +", paste(covariates, collapse = " + "), "+ KRT57negative"))

multi_model <- coxph(multi_formula, data = Cox_df)

multi_res <- summary(multi_model)

plot_df_prop <- data.frame(group = rownames(multi_res$coefficients),
                      p_value = signif(multi_res$coefficients[,"Pr(>|z|)"],2),
                      HR = signif(multi_res$coefficients[,"exp(coef)"],2),
                      CI_low = signif(multi_res$conf.int[,"lower .95"],2),
                      CI_high = signif(multi_res$conf.int[,"upper .95"],2),
                      wald_test = signif(multi_res$waldtest["pvalue"],2))


#Plot hazard ratios
hr_prop <- ggplot(plot_df_prop, aes(x = reorder(group,desc(HR)), y = HR, color = p_value < 0.05))+
geom_point(size = 4)+
geom_errorbar(aes(ymax = CI_high, ymin = CI_low))+
geom_hline(yintercept = 1,color = "black",linetype = "dashed")+
scale_color_brewer(palette = "Set1")+
xlab("Covariates")+
ylab("HR for PID groups")+
theme_classic()+
theme(axis.text.x = element_text(angle = 90))

hr_prop
```

## Barplot visualization 

```{r}
dittoBarPlot(vsd, var = "KRT5_7_score", group.by = "TNBC")+
    scale_fill_manual(values = metadata(vsd)$colors$KRT57_score)
dittoBarPlot(vsd, var = "KRT5_7_score", group.by = "TNBC", scale = "count")+
  scale_fill_manual(values = metadata(vsd)$colors$KRT57_score)

dittoBarPlot(vsd, var = "KRT5_7_score", group.by = "PAM50Call_RNAseq")+
    scale_fill_manual(values = metadata(vsd)$colors$KRT57_score)
dittoBarPlot(vsd, var = "KRT5_7_score", group.by = "PAM50Call_RNAseq", scale = "count")+
      scale_fill_manual(values = metadata(vsd)$colors$KRT57_score)

dittoBarPlot(vsd, var = "PAM50Call_RNAseq", group.by = "KRT5_7_score")+
        scale_fill_manual(values = metadata(vsd)$colors$PAM50)
dittoBarPlot(vsd, var = "PAM50Call_RNAseq", group.by = "KRT5_7_score", scale = "count")+
  scale_fill_manual(values = metadata(vsd)$colors$PAM50)

dittoBarPlot(vsd, var = "TNBC", group.by = "KRT5_7_score")
dittoBarPlot(vsd, var = "TNBC", group.by = "KRT5_7_score", scale = "count")
```

## Differential expression analysis 

### 1. Perform DEA based on combined score 

```{r}
brca.clean$KRT5_7_score <- vsd$KRT5_7_score[match(brca.clean$PID, vsd$PID)]
brca.clean$KRT5_7_score_combined <- ifelse(brca.clean$KRT5_7_score == "KRT5/7+","KRT5_7_pos","Not_KRT5_7_pos")
```

```{r}
saveRDS(brca.clean,"/mnt/rcc_volume/TNBC/data_analysis/TCGA_BRCA_cleaned_full.rds")
brca.clean <- readRDS("/mnt/rcc_volume/TNBC/data_analysis/TCGA_BRCA_cleaned_full.rds")
```

```{r}
library(DESeq2)

# Create DESeqDataset
ddsSE_1 <- DESeqDataSet(brca.clean, design = ~ KRT5_7_score_combined)

# Filter genes
summary(rowSums(counts(ddsSE_1)) >= 10)
keep <- rowSums(counts(ddsSE_1)) >= 10
ddsSE_1 <- ddsSE_1[keep, ]


# DESeq analysis includes count normalization based on size factors
dds <- DESeq(ddsSE_1, )

# Extract results - also performs independent filtering
res <- DESeq2::results(dds) #alpha = 0.05, lfcThreshold = 1)
summary(res)

# Variance-stabilizing transformation
vsd <- DESeq2::vst(dds)
```

```{r}
#Save objects 
saveRDS(res, "/mnt/rcc_volume/TNBC/data_analysis/TCGA_BRCA_DESeq_VST_full_CombinedScore.rds")
saveRDS(vsd, "/mnt/rcc_volume/TNBC/data_analysis/TCGA_BRCA_DESeq_VST_full_Combined_VST.rds")

res <- readRDS("/mnt/rcc_volume/TNBC/data_analysis/TCGA_BRCA_DESeq_VST_full_DiffResults.rds")
vsd <- readRDS("/mnt/rcc_volume/TNBC/data_analysis/TCGA_BRCA_DESeq_VST_full_DiffResults_VST.rds")
```

#### Heatmap visualization 

```{r}
# Extract Top 20
rownames(res) <- rowData(brca.clean)$gene_name[match(rownames(res), rownames(brca.clean))]

sig.up <- res %>% as.data.frame %>% filter(log2FoldChange > 1) %>% arrange(padj) %>% slice(1:30) %>% rownames_to_column("gene") %>% pull(gene)
sig.down <- res %>% as.data.frame %>% filter(log2FoldChange < -1) %>% arrange(padj) %>% slice(1:30) %>% rownames_to_column("gene") %>% pull(gene)

# Heatmap body 
ht <- assay(vsd) %>% as.data.frame %>% rownames_to_column("gene")
ht$gene <- rowData(brca.clean)$gene_name[match(ht$gene, rownames(brca.clean))]
mat_up <- ht %>% filter(gene %in% sig.up) %>% column_to_rownames("gene")
mat_down <- ht %>% filter(gene %in% sig.down) %>% column_to_rownames("gene")

# Annotation
anno <- as.data.frame(t(mat_up))
anno$KRT5_7_score_combined <- brca.clean$KRT5_7_score_combined[match(rownames(anno),colnames(brca.clean))]
anno$KRT5_7_score <- brca.clean$KRT5_7_score[match(rownames(anno),colnames(brca.clean))]
anno$PAM50 <- brca.clean$PAM50Call_RNAseq[match(rownames(anno),colnames(brca.clean))]
anno$TNBC <- brca.clean$TNBC[match(rownames(anno),colnames(brca.clean))]

# Plot Heatmap
Heatmap(mat_up, 
        col = viridis(100),
        column_split = anno$KRT5_7_score_combined, 
        column_title = NULL)%v%
  columnAnnotation(KRT5_7_score_combined = anno$KRT5_7_score_combined)%v%
  columnAnnotation(KRT5_7_score = anno$KRT5_7_score)%v%
  columnAnnotation(PAM50 = anno$PAM50)%v%
  columnAnnotation(TNBC = anno$TNBC)

Heatmap(mat_down, 
        col = viridis(100),
        column_split = anno$KRT5_7_score_combined, 
        column_title = NULL)%v%
  columnAnnotation(KRT5_7_score_combined = anno$KRT5_7_score_combined)%v%
  columnAnnotation(KRT5_7_score = anno$KRT5_7_score)%v%
  columnAnnotation(PAM50 = anno$PAM50)%v%
  columnAnnotation(TNBC = anno$TNBC)
```

#### Volcano-plot visualization 

```{r}
## LFC shrinkage using apeglm 
DESeq2::plotMA(res, ylim = c(-5, 5))

plot_vol <- res %>% as.data.frame %>% rownames_to_column("gene")

library(apeglm)
DESeq2::resultsNames(dds)

reshape <- DESeq2::lfcShrink(dds, coef = "KRT5_7_score_combined_Not_KRT5_7_pos_vs_KRT5_7_pos", 
                            type = "apeglm")

DESeq2::plotMA(reshape, ylim = c(-5, 5))
```

```{r}
#rownames(reshape) <- rowData(brca.clean)$gene_name[match(rownames(reshape), rownames(brca.clean))]

plot_vol <- res %>% as.data.frame %>% rownames_to_column("gene")
plot_vol <- plot_vol %>% mutate(type = ifelse(log2FoldChange > 1 & padj < 0.05, "UP", 
                                  ifelse(log2FoldChange < -1 & padj < 0.05, "DOWN","NA")))

col <- c("DOWN" = "#E41A1C", "UP" = "#377EB8", "NA" = "black")

# Down-regulated
ggplot(plot_vol, aes(x = log2FoldChange, y=-log10(padj)))+
  geom_point(aes(color = type))+
  geom_label_repel(data = plot_vol %>% filter(gene %in% sig.down), aes(label = gene, color = type), max.overlaps = 30)+
  xlab("log2FC (CK5/7+ vs Not_CK5/7+)")+
  ylab("-log10(FDR)")+
  geom_hline(yintercept = -log10(0.01), linetype = "dashed")+
  geom_hline(yintercept = -log10(1), linetype = "dashed")+
  geom_vline(xintercept = 1, linetype = "dashed")+
  geom_vline(xintercept = -1, linetype = "dashed")+
  theme_classic()+
  scale_color_manual(values = col)+
  guides(color = "none")


# Up-regulated
ggplot(plot_vol, aes(x = log2FoldChange, y=-log10(padj)))+
  geom_point(aes(color = type))+
  geom_label_repel(data = plot_vol %>% filter(gene %in% sig.up), aes(label = gene, color = type), max.overlaps = 30)+
  xlab("log2FC (CK5/7+ vs Not_CK5/7+)")+
  ylab("-log10(FDR)")+
  geom_hline(yintercept = -log10(0.01), linetype = "dashed")+
  geom_hline(yintercept = -log10(1), linetype = "dashed")+
  geom_vline(xintercept = 1, linetype = "dashed")+
  geom_vline(xintercept = -1, linetype = "dashed")+
  theme_classic()+
  scale_color_manual(values = col)+
  guides(color = "none")
```

#### Boxplot visualization for GOI 

```{r}
p1 <- ggplot(anno, aes(x = KRT5_7_score_combined, y = TBC1D9))+
  geom_boxplot(outlier.colour = NA)+
  geom_jitter(aes(color = KRT5_7_score_combined), size = 3, position=position_jitter(w=0.1,h=0.1))+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90))+
  xlab("")+
  #stat_compare_means(method = "kruskal.test")+
  scale_color_brewer(palette = "Set1")+
  guides(color = "none")

p2 <- ggplot(anno, aes(x = KRT5_7_score, y = TBC1D9))+
  geom_boxplot(outlier.colour = NA)+
  geom_jitter(aes(color = KRT5_7_score_combined), size = 3, position=position_jitter(w=0.1,h=0.1))+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90))+
  ylab("")+
  xlab("")+
  scale_color_brewer(palette = "Set1")+
  guides(color = "none")
  
p3 <- ggplot(anno, aes(x = TNBC, y = TBC1D9))+
  geom_boxplot(outlier.colour = NA)+
  geom_jitter(aes(color = KRT5_7_score_combined), size = 3, position=position_jitter(w=0.1,h=0.1))+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90))+
  ylab("")+
  xlab("")+
  scale_color_brewer(palette = "Set1")+
  guides(color = "none")


p1+p2+p3
```

### 2. Perform DEA based on all scores

```{r}
brca.clean$KRT5_7_score <- str_replace(brca.clean$KRT5_7_score, "\\+","_pos")
brca.clean$KRT5_7_score <- str_replace(brca.clean$KRT5_7_score, "\\-","_neg")
brca.clean$KRT5_7_score <- factor(brca.clean$KRT5_7_score)
brca.clean$KRT5_7_score <- relevel(brca.clean$KRT5_7_score, ref = "KRT5/7_pos")

library(ExploreModelMatrix)
vd <- VisualizeDesign(sampleData = colData(brca.clean), 
                      designFormula = ~ KRT5_7_score)
vd$plotlist
```

```{r}
library(DESeq2)

# relevel KRT5_7_score 
# perform DESeq 

# Create DESeqDataset
ddsSE_1 <- DESeqDataSet(brca.clean, design = ~ KRT5_7_score)

# DESeq analysis includes count normalization based on size factors
dds <- DESeq(ddsSE_1)

# Extract results - also performs independent filtering
res <- DESeq2::results(dds)
summary(res)

# Variance-stabilizing transformation
vsd <- DESeq2::vst(dds)
```

```{r}
#Save objects 
saveRDS(res, "/mnt/rcc_volume/TNBC/data_analysis/TCGA_BRCA_DESeq_VST_full_FullScore.rds")
saveRDS(vsd, "/mnt/rcc_volume/TNBC/data_analysis/TCGA_BRCA_DESeq_VST_full_FullScore_VST.rds")

#res <- readRDS("/mnt/rcc_volume/TNBC/data_analysis/TCGA_BRCA_DESeq_VST_full_DiffResults.rds")
#vsd <- readRDS("/mnt/rcc_volume/TNBC/data_analysis/TCGA_BRCA_DESeq_VST_full_DiffResults_VST.rds")
```

### Gene-set enrichment analysis 

With `clusterProfiler`. 

```{r}
library(clusterProfiler)
library(enrichplot)
library(org.Hs.eg.db)

organism <- "org.Hs.eg.db"

# Prepare input gene list based on log2foldchanges 
original_gene_list <- res$log2FoldChange
names(original_gene_list) <- rownames(res)
gene_list <- na.omit(original_gene_list)
gene_list <- sort(gene_list, decreasing = TRUE) # sort in decreasing order

# Run GO analysis
gse <- gseGO(geneList=gene_list, 
             ont ="BP", 
             keyType = "SYMBOL", 
             minGSSize = 3, 
             maxGSSize = 1000, 
             pvalueCutoff = 0.05, 
             verbose = TRUE, 
             OrgDb = organism, 
             pAdjustMethod = "BH")

# Visualization
## Dot plot
dotplot(gse, showCategory=10, split=".sign") + facet_grid(~.sign)

gse %>% as.data.frame()

## Emap plot
emapplot(pairwise_termsim(gse), showCategory = 20, color = "enrichmentScore", layout = "circle")

## GSEA plot
gseaplot2(gse, title = gse$Description[11], geneSetID = 1)
gseaplot2(gse, title = gse$Description[9], geneSetID = 1)
```

With `goana`. Over-representation of genes in pathways.

```{r}
# goana() requires Entrez IDs, some of which map to multiple
# symbols - hence the unique() in the call below.
library(org.Hs.eg.db)
entrez.ids <- mapIds(org.Hs.eg.db, keys=rownames(res), 
                     column="ENTREZID", keytype="SYMBOL")

sig.down <- res %>% as.data.frame %>% filter(log2FoldChange < -1 & padj < 0.01) %>% rownames_to_column("gene") %>% pull("gene")

entrez.ids.DE <- mapIds(org.Hs.eg.db, keys=sig.down, 
                     column="ENTREZID", keytype="SYMBOL")

library(limma)
go.out <- goana(unique(entrez.ids.DE), species="Hs", 
                universe=unique(entrez.ids))


# Only keeping biological process terms that are not overly general
go.out <- go.out[order(go.out$P.DE),]
go.useful <- go.out[go.out$Ont=="BP" & go.out$N <= 2000,]

plot_go_BP <- go.useful[1:10,] #Top10 Genes

ggplot(plot_go_BP, aes(x = reorder(Term,-P.DE), y = -log10(P.DE)))+
  geom_bar(stat = "identity", fill = "firebrick", color = "black")+
  coord_flip()+
  theme_classic()
```

## KEGG pathway enrichment 

With `clusterProfiler`. 

```{r}
# KEGG requires Entrez IDs 
library(org.Hs.eg.db)

entrez.ids <- mapIds(org.Hs.eg.db, keys=rownames(res), column="ENTREZID", keytype="SYMBOL")
entrez.ids <- entrez.ids[!duplicated(entrez.ids)]

df <- data.frame(log_change = gene_list, 
           SYMBOL = names(gene_list))

df$ENTREZ <- entrez.ids[match(df$SYMBOL, names(entrez.ids))]
df <- df[!is.na(df$ENTREZ),]

gene_list_kegg <- df$log_change
names(gene_list_kegg) <- df$ENTREZ

# Run KEGG
gseKEGG <- gseKEGG(geneList=gene_list_kegg, 
        organism = "hsa",
        keyType = "kegg", 
        minGSSize = 10, 
        maxGSSize = 500, 
        pvalueCutoff = 0.05, 
        verbose = TRUE, 
        pAdjustMethod = "BH")

# Visualization 
## Dotplot 
dotplot(gseKEGG, showCategory = 15, title = "Enriched Pathways" , split=".sign") + facet_grid(.~.sign)

## EMAP plot 
emapplot(pairwise_termsim(gseKEGG), showCategory = 15,color = "enrichmentScore", layout = "circle")

## GSEA plot
gseaplot2(gseKEGG, title = gseKEGG$Description[2], geneSetID = 1)
```


## Immune cell deconvolution 

```{r}
library(immunedeconv)

deconvolution_methods

# TPM gene expression matrix 
TPM_mat <- assay(brca.clean, "tpm_unstrand")
rownames(TPM_mat) <- rowData(brca.clean)$gene_name
```

### MCP-counter 

```{r}
res_MCP <- readRDS("/mnt/rcc_volume/TNBC/data_analysis/TCGA_BRCA_MCP.rds")
#res_MCP <- immunedeconv::deconvolute(TPM_mat, method = "mcp_counter")

#Z-scoring of MCP results by cell type
res_MCP_1 <- res_MCP %>% dplyr::select(-cell_type)
rownames(res_MCP_1) <- res_MCP$cell_type
res_MCP_z <- as.data.frame(t(scale(t(res_MCP_1), center=TRUE, scale=TRUE)))
res_MCP_z <- res_MCP_z %>% rownames_to_column("cell_type")

res_MCP_plot <- res_MCP_z %>% pivot_longer(2:ncol(res_MCP_z), names_to = "PID", values_to = "score")

res_MCP_plot$KRT5_7_score_combined <- brca.clean$KRT5_7_score_combined[match(res_MCP_plot$PID, colnames(brca.clean))]
res_MCP_plot$KRT5_7_score <- brca.clean$KRT5_7_score[match(res_MCP_plot$PID, colnames(brca.clean))]
res_MCP_plot$TNBC <- brca.clean$TNBC[match(res_MCP_plot$PID, colnames(brca.clean))]

ggplot(res_MCP_plot, aes(x = KRT5_7_score, y = score))+
  geom_boxplot(outlier.colour = NA)+
  geom_jitter(aes(color = KRT5_7_score), size = 3, position=position_jitter(w=0.1,h=0.1))+
  facet_wrap(~cell_type, scales = "free_y")+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90))+
  scale_color_brewer(palette = "Set1")+
  guides(color = "none")+
  scale_color_manual(values = metadata(vsd)$colors$KRT57_score)
  #stat_compare_means(method = "t.test")
  

# Heatmap visualization  
ht_MCP <- res_MCP_plot %>% group_by(KRT5_7_score, cell_type) %>% summarize(mean = mean(score)) %>% pivot_wider(names_from = KRT5_7_score, values_from = mean) %>% column_to_rownames("cell_type")

Heatmap(ht_MCP, name = "mean MCP-Z-score")
```

### Quantiseq 

```{r}
res_QS <- readRDS("/mnt/rcc_volume/TNBC/data_analysis/TCGA_BRCA_Quantiseq.rds")
#res_QS <- immunedeconv::deconvolute(TPM_mat, method = "quantiseq", tumor = TRUE)

plot_QS <- res_QS %>% pivot_longer(2:ncol(res_QS), names_to = "PID", values_to = "fra")
plot_QS$KRT5_7_score_combined <- brca.clean$KRT5_7_score_combined[match(plot_QS$PID, colnames(brca.clean))]
plot_QS$KRT5_7_score <- brca.clean$KRT5_7_score[match(plot_QS$PID, colnames(brca.clean))]
plot_QS$TNBC <- brca.clean$TNBC[match(plot_QS$PID, colnames(brca.clean))]


ggplot(plot_QS, aes(x = KRT5_7_score, y = fra))+
  geom_boxplot(outlier.colour = NA, aes(color = KRT5_7_score))+
  #geom_jitter(aes(color = KRT5_7_score), size = 3, position=position_jitter(w=0.1,h=0.1))+
  facet_wrap(~cell_type, scales = "free_y")+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90))+
  ylab("")+
  xlab("")+
  scale_color_manual(values = metadata(vsd)$colors$KRT57_score)+
  guides(color = "none")
  #stat_compare_means(method = "t.test")

ggplot(plot_QS, aes(x = TNBC, y = fra))+
  geom_boxplot(outlier.colour = NA,aes(color = TNBC))+
  #geom_jitter(aes(color = KRT5_7_score), size = 3, position=position_jitter(w=0.1,h=0.1))+
  facet_wrap(~cell_type, scales = "free_y")+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90))+
  ylab("")+
  xlab("")+
  scale_color_brewer(palette = "Set1")+
  guides(color = "none")

# Heatmap visualization  
ht_QS <- plot_QS %>% group_by(KRT5_7_score, cell_type) %>% summarize(mean = mean(fra)) %>% pivot_wider(names_from = KRT5_7_score, values_from = mean) %>% column_to_rownames("cell_type")

Heatmap(ht_QS, name = "mean QS-fraction", col = viridis(100))
Heatmap(t(scale(t(ht_QS))), name = "mean QS-fraction")
```

```{r}
saveRDS(res_MCP, "/mnt/rcc_volume/TNBC/data_analysis/TCGA_BRCA_MCP.rds")
saveRDS(res_QS, "/mnt/rcc_volume/TNBC/data_analysis/TCGA_BRCA_Quantiseq.rds")
```


